core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -1153,39 +1153,29 @@ private[sql] final class PreprocessTableInsertOrPut(conf: SQLConf)
     // ResolveRelations, no such special rule has been added for PUT
     case d@DeleteFromTable(table, child) if table.resolved && child.resolved =>
       EliminateSubqueryAliases(table) match {
-        case l@LogicalRelation(dr: DeletableRelation, _, _) =>
+        case l@LogicalRelation(dr: MutableRelation, _, _) =>
           def comp(a: Attribute, targetCol: String): Boolean = a match {
             case ref: AttributeReference => targetCol.equals(ref.name.toUpperCase)
           }
 
-          val expectedOutput = l.output
-          if (!child.output.forall(a => expectedOutput.exists(e => comp(a, e.name.toUpperCase)))) {
-            throw new AnalysisException(s"$l requires that the query in the " +
-                "WHERE clause of the DELETE FROM statement " +
-                "generates the same column name(s) as in its schema but found " +
-                s"${child.output.mkString(",")} instead.")
-          }
-          l match {
-            case LogicalRelation(ps: PartitionedDataSourceScan, _, _) =>
-              if (!ps.partitionColumns.forall(a => child.output.exists(e =>
-                comp(e, a.toUpperCase)))) {
-                throw new AnalysisException(s"${child.output.mkString(",")}" +
-                    s" columns in the WHERE clause of the DELETE FROM statement must " +
-                    s"have all the parititioning column(s) ${ps.partitionColumns.mkString(",")}.")
-              }
-            case _ =>
+          val childAttributes = child.output.map(ar => (ar.name.toUpperCase(), ar)).toMap
+          val keyColumns = dr.getPrimaryKeyColumns
+
+          val childOutput = try {
+            keyColumns.map(_.toUpperCase)
+                .map(childAttributes(_))
+          } catch {
+            case _: NoSuchElementException =>
+              throw new AnalysisException(s"$l requires that the query in the " +
+                  "WHERE clause of the DELETE FROM statement " +
+                  s"must have all the key column(s) ${keyColumns.mkString(",")} but found " +
+                  s"${child.output.mkString(",")} instead.")
```

Comment:
Will be cleaner to match instead of catching exception like:

```scala
val childOutput = keyColumns.map(col => childAttributes.get(col.toUpperCase) match {
  case Some(a) => a
  case None => throw new AnalysisException(...)
})
```

Better way is to use resolveQuoted -- see comment below.